modules/tests/model_based.rs

```diff
@@ -0,0 +1,266 @@
+use ibc::ics26_routing::msgs::ICS26Envelope;
+use ibc::mock::client_state::{MockClientState, MockConsensusState};
+use ibc::mock::context::MockContext;
+use ibc::mock::header::MockHeader;
+use ibc::mock::host::HostType;
+use ibc::Height;
+use std::collections::HashMap;
+use std::error::Error;
+use std::fmt::{Debug, Display};
+use step::{ActionOutcome, ActionType, Chain, Step};
+use tendermint::account::Id as AccountId;
+
+#[derive(Debug)]
+struct IBCTestExecutor {
+    // mapping from chain identifier to its context
+    contexts: HashMap<String, MockContext>,
+}
+
+impl IBCTestExecutor {
+    fn new() -> Self {
+        Self {
+            contexts: Default::default(),
+        }
+    }
+
+    /// Create a `MockContext` for a given `chain_id`.
+    /// Panic if a context for `chain_id` already exists.
+    fn init_chain_context(&mut self, chain_id: String, initial_height: u64) {
+        let max_history_size = 1;
+        let ctx = MockContext::new(
+            ChainId::new(chain_id.clone(), Self::epoch()),
+            HostType::Mock,
+            max_history_size,
+            Height::new(Self::epoch(), initial_height),
+        );
+        assert!(self.contexts.insert(chain_id, ctx).is_none());
+    }
+
+    /// Returns a reference to the `MockContext` of a given `chain_id`.
+    /// Panic if the context for `chain_id` is not found.
+    fn chain_context(&self, chain_id: &String) -> &MockContext {
+        self.contexts
+            .get(chain_id)
+            .expect("chain context should have been initialized")
+    }
+
+    /// Returns a mutable reference to the `MockContext` of a given `chain_id`.
+    /// Panic if the context for `chain_id` is not found.
+    fn chain_context_mut(&mut self, chain_id: &String) -> &mut MockContext {
+        self.contexts
+            .get_mut(chain_id)
+            .expect("chain context should have been initialized")
+    }
+
+    fn extract_handler_error_kind<K>(ics18_result: Result<(), ICS18Error>) -> K
+    where
+        K: Clone + Debug + Display + Into<anomaly::BoxError> + 'static,
+    {
+        let ics18_error = ics18_result.expect_err("ICS18 error expected");
+        assert!(matches!(
+            ics18_error.kind(),
+            ICS18ErrorKind::TransactionFailed
+        ));
+        let ics26_error = ics18_error
+            .source()
+            .expect("expected source in ICS18 error")
+            .downcast_ref::<ICS26Error>()
+            .expect("ICS18 source should be an ICS26 error");
+        assert!(matches!(
+            ics26_error.kind(),
+            ICS26ErrorKind::HandlerRaisedError,
+        ));
+        ics26_error
+            .source()
+            .expect("expected source in ICS26 error")
+            .downcast_ref::<anomaly::Error<K>>()
+            .expect("ICS26 source should be an handler error")
+            .kind()
+            .clone()
+    }
+
+    // TODO: this is sometimes called version/revision number but seems
```

Comment:
Chain "epoch" was redefined to "revision" I think, we just didn't get a chance to make it consistent thoughout the codebase. The same goes for the epoch in heights. Ref: https://github.com/informalsystems/ibc-rs/issues/405

True this has nothing to do with `Version` type for channels or connections.

Reply:
Thanks. Renamed `epoch` to `revision` in 75db6bb9621e3a6db269100332e137d3a526a822.